pairing/bn254/point.go

```diff
@@ -0,0 +1,723 @@
+		p.g.y = *newGFp(1)
+		p.g.z = gfP{0}
+		p.g.t = gfP{0}
+	} else {
+		p.g.z = *newGFp(1)
+		p.g.t = *newGFp(1)
+	}
+
+	if !p.g.IsOnCurve() {
+		return errors.New("bn254.G1: malformed point")
+	}
+
+	return nil
+}
+
+func (p *pointG1) UnmarshalFrom(r io.Reader) (int, error) {
+	buf := make([]byte, p.MarshalSize())
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		return n, err
+	}
+	return n, p.UnmarshalBinary(buf)
+}
+
+func (p *pointG1) MarshalSize() int {
+	return 2 * p.ElementSize()
+}
+
+func (p *pointG1) ElementSize() int {
+	return 256 / 8
+}
+
+func (p *pointG1) String() string {
+	return "bn254.G1" + p.g.String()
+}
+
+func (p *pointG1) Hash(m []byte) kyber.Point {
+	return hashToPoint(p.dst, m)
+}
+
+func hashToPoint(domain, m []byte) kyber.Point {
+	e0, e1 := hashToField(domain, m)
+	p0 := mapToPoint(domain, e0)
+	p1 := mapToPoint(domain, e1)
+	p := p0.Add(p0, p1)
+	return p
+}
+
+func hashToField(domain, m []byte) (*gfP, *gfP) {
+	const u = 48
+	_msg := expandMsgXmdKeccak256(domain, m, 2*u)
+	x, y := new(big.Int), new(big.Int)
+	x.SetBytes(_msg[0:48]).Mod(x, p)
+	y.SetBytes(_msg[48:96]).Mod(y, p)
+	gx, gy := &gfP{}, &gfP{}
+	gx.Unmarshal(zeroPadBytes(x.Bytes(), 32))
+	gy.Unmarshal(zeroPadBytes(y.Bytes(), 32))
+	montEncode(gx, gx)
+	montEncode(gy, gy)
+	return gx, gy
+}
+
+// `mapToPoint` implements the general Shallue-van de Woestijne mapping to BN254 G1
+// RFC9380, 6.6.1. https://datatracker.ietf.org/doc/html/rfc9380#name-shallue-van-de-woestijne-me
+func mapToPoint(domain []byte, u *gfP) kyber.Point {
+	tv1 := &gfP{}
+	tv1.Set(u)
+	gfpMul(tv1, tv1, tv1)
+	gfpMul(tv1, tv1, c1)
+	tv2 := &gfP{}
+	gfpAdd(tv2, newGFp(1), tv1)
+	negTv1 := &gfP{}
+	gfpNeg(negTv1, tv1)
+	gfpAdd(tv1, newGFp(1), negTv1)
+	tv3 := &gfP{}
+	gfpMul(tv3, tv1, tv2)
+	tv3.Invert(tv3)
+	tv5 := &gfP{}
+	gfpMul(tv5, u, tv1)
+	gfpMul(tv5, tv5, tv3)
+	gfpMul(tv5, tv5, c3)
+	x1 := &gfP{}
+	gfpSub(x1, c2, tv5)
+	x2 := &gfP{}
+	gfpAdd(x2, c2, tv5)
+	tv7 := &gfP{}
+	gfpMul(tv7, tv2, tv2)
+	tv8 := &gfP{}
+	gfpMul(tv8, tv7, tv3)
+	x3 := &gfP{}
+	gfpMul(x3, tv8, tv8)
+	gfpMul(x3, c4, x3)
+	gfpAdd(x3, newGFp(1), x3)
+
+	x, y := &gfP{}, &gfP{}
+	if legendre(g(x1)) == 1 {
+		x = x1
+		y.Sqrt(g(x1))
+	} else if legendre(g(x2)) == 1 {
+		x = x2
+		y.Sqrt(g(x2))
+	} else {
+		x = x3
+		y.Sqrt(g(x3))
+	}
+	if sgn0(u) != sgn0(y) {
+		gfpNeg(y, y)
+	}
+
+	p := newPointG1(domain).Base().(*pointG1)
+	p.g.x.Set(x)
+	p.g.y.Set(y)
+	return p
+}
+
+// `expandMsgXmdKeccak256` implements expand_message_xmd from IETF RFC9380 Sec 5.3.1
+// Borrowed from: https://github.com/kilic/bls12-381/blob/master/hash_to_field.go
+func expandMsgXmdKeccak256(domain, msg []byte, outLen int) []byte {
+	h := sha3.NewLegacyKeccak256()
+	domainLen := uint8(len(domain))
+	if domainLen > 255 {
+		panic("invalid domain length")
+	}
+	// DST_prime = DST || I2OSP(len(DST), 1)
+	// b_0 = H(Z_pad || msg || l_i_b_str || I2OSP(0, 1) || DST_prime)
+	_, _ = h.Write(make([]byte, h.BlockSize()))
```

Comment:
in the `gnarkExpandMsgXmd` we check/return errors but here we don't. Is there a reason to swallow them here? Admittedly they're unlikely to happen, but it seems like good hygiene

Reply:
The `HashablePoint.Hash` doesn't have an error in its return, I tried to handle the errors but the cascading effect goes out of hand. This pr might not be the place to do that ?